src/main/java/picard/sam/util/SamComparison.java

```diff
@@ -39,20 +51,109 @@
+        this.ignoreSpecies = ignoreSpecies;
+        this.ignoreAssembly = ignoreAssembly;
+        this.ignoreUR = ignoreUR;
+        this.ignoreM5 = ignoreM5;
+        this.ignorePG = ignorePG;
+        this.lenientDup = lenientDup;
+        this.mq0Match = mq0Match;
+        this.leftName = leftName;
+        this.rightName = rightName;
+        comparisonMetric = new SamComparisonMetric();
+        setup();
         this.headersAreEqual = compareHeaders();
         this.alignmentsAreEqual = compareAlignments();
+        this.duplicateMarksAreEqual = compareDuplicateMarks();
+        comparisonMetric.areEqual = areEqual();
     }
 
-    public void printReport() {
-        System.out.println("Match\t" + mappingsMatch);
-        System.out.println("Differ\t" + mappingsDiffer);
-        System.out.println("Unmapped_both\t" + unmappedBoth);
-        System.out.println("Unmapped_left\t" + unmappedLeft);
-        System.out.println("Unmapped_right\t" + unmappedRight);
-        System.out.println("Missing_left\t" + missingLeft);
-        System.out.println("Missing_right\t" + missingRight);
+    public void writeReport(final File output) {
+        writeReport(output, new ArrayList<>());
+    }
+
+    public void writeReport(final File output, final List<Header> headers) {
+        final MetricsFile<SamComparisonMetric, ?> comparisonMetricFile = new MetricsFile<>();
+        for (final Header h : headers) {
+            comparisonMetricFile.addHeader(h);
+        }
+        comparisonMetricFile.addAllMetrics(Collections.singletonList(comparisonMetric));
+        comparisonMetricFile.write(output);
+    }
+
+    private void setup() {
+        if (lenientDup) {
+            /*Setup for duplicate marking checks.  Recs which disagree on duplicate marking will be added to markDuplicatesCheckLeft/Right and the fed to a DuplicateSetIterator to check if
+             * differences are due only to choice of representative read within duplicate set.
+             */
+            final SAMFileHeader leftHeader = reader1.getFileHeader();
+            final SAMFileHeader rightHeader = reader2.getFileHeader();
+            final SAMRecordDuplicateComparator leftDupComparator = new SAMRecordDuplicateComparator(Collections.singletonList(leftHeader));
+            final SAMRecordDuplicateComparator rightDupComparator = new SAMRecordDuplicateComparator(Collections.singletonList(rightHeader));
+            markDuplicatesCheckLeft = SortingCollection.newInstance(SAMRecord.class, new BAMRecordCodec(leftHeader),
+                    leftDupComparator, SAMFileWriterImpl.getDefaultMaxRecordsInRam());
+            markDuplicatesCheckRight = SortingCollection.newInstance(SAMRecord.class, new BAMRecordCodec(rightHeader),
+                    rightDupComparator, SAMFileWriterImpl.getDefaultMaxRecordsInRam());
+        }
+        comparisonMetric.leftFile = leftName;
+        comparisonMetric.rightFile = rightName;
+    }
+
+    private boolean compareDuplicateMarks() {
```

Comment:
Its quite hard to see the implicit dependencies embedded in these various methods. Most of the code in this method is about updating `duplicateMarkingsDiffer`, but only if its in `lenientDup` mode, and that code is entirely dependent on side effects of `compareAlignments` (or actually, side effects of the various methods called by `compareAlignments` - ultimately I think they funnel down into catalogDuplicateDifferences` ?).

A little refactoring would make it much easier to see, or even just comments documenting the dependencies would help a lot. I'd also rename this method to something like `getUpdateDuplicateMarkingsDiffer` or `resolveDuplicateMarkings`, and add a comment to `compareAlignments` documenting that it's responsible for updating the global duplicate set iterators.

Reply:
I agree, this is organized a bit confusingly.  I've done some refactoring and added some documentation which has hopefully cleared things up a bit.